File: backend/app/permissions.py
"""
Permission system for role-based access control.
"""

from enum import Enum
from typing import Dict, List, Set
from .models.user import UserRole


class Permission(Enum):
    """
    Enumeration of all permissions in the system.
    Each permission represents a specific action that can be performed.
    """
    
    # User Management Permissions
    CREATE_USER = "create_user"
    READ_USER = "read_user"
    READ_ALL_USERS = "read_all_users"
    UPDATE_USER = "update_user"
    UPDATE_ANY_USER = "update_any_user"
    DELETE_USER = "delete_user"
    DELETE_ANY_USER = "delete_any_user"
    ASSIGN_ROLE = "assign_role"
    SUSPEND_USER = "suspend_user"
    
    # Course Management Permissions
    CREATE_COURSE = "create_course"
    READ_COURSE = "read_course"
    READ_ALL_COURSES = "read_all_courses"
    UPDATE_COURSE = "update_course"
    UPDATE_ANY_COURSE = "update_any_course"
    DELETE_COURSE = "delete_course"
    DELETE_ANY_COURSE = "delete_any_course"
    PUBLISH_COURSE = "publish_course"
    FEATURE_COURSE = "feature_course"
    HIDE_COURSE = "hide_course"
    
    # Content Management Permissions
    UPLOAD_CONTENT = "upload_content"
    UPDATE_CONTENT = "update_content"
    DELETE_CONTENT = "delete_content"
    MODERATE_CONTENT = "moderate_content"
    
    # Enrollment Permissions
    ENROLL_COURSE = "enroll_course"
    VIEW_ENROLLMENTS = "view_enrollments"
    VIEW_ALL_ENROLLMENTS = "view_all_enrollments"
    MANAGE_ENROLLMENTS = "manage_enrollments"
    
    # Learning Permissions
    VIEW_PROGRESS = "view_progress"
    VIEW_ALL_PROGRESS = "view_all_progress"
    UPDATE_PROGRESS = "update_progress"
    TAKE_QUIZ = "take_quiz"
    CREATE_NOTE = "create_note"
    
    # Q&A and Discussion Permissions
    ASK_QUESTION = "ask_question"
    ANSWER_QUESTION = "answer_question"
    MODERATE_QA = "moderate_qa"
    
    # Analytics and Reporting Permissions
    VIEW_ANALYTICS = "view_analytics"
    VIEW_OWN_ANALYTICS = "view_own_analytics"
    VIEW_COURSE_ANALYTICS = "view_course_analytics"
    VIEW_PLATFORM_ANALYTICS = "view_platform_analytics"
    EXPORT_ANALYTICS = "export_analytics"
    
    # Financial Permissions
    PROCESS_PAYMENTS = "process_payments"
    VIEW_TRANSACTIONS = "view_transactions"
    VIEW_ALL_TRANSACTIONS = "view_all_transactions"
    MANAGE_PAYOUTS = "manage_payouts"
    VIEW_REVENUE = "view_revenue"
    
    # System Administration Permissions
    MANAGE_CATEGORIES = "manage_categories"
    MANAGE_SETTINGS = "manage_settings"
    MANAGE_EMAIL_TEMPLATES = "manage_email_templates"
    VIEW_SYSTEM_HEALTH = "view_system_health"
    MANAGE_LEGAL_PAGES = "manage_legal_pages"
    
    # Certificate Permissions
    ISSUE_CERTIFICATE = "issue_certificate"
    VIEW_CERTIFICATES = "view_certificates"
    VERIFY_CERTIFICATE = "verify_certificate"


# Role-Permission Mapping
ROLE_PERMISSIONS: Dict[UserRole, Set[Permission]] = {
    UserRole.SUPER_ADMIN: {
        # Super Admin has all permissions
        Permission.CREATE_USER,
        Permission.READ_USER,
        Permission.READ_ALL_USERS,
        Permission.UPDATE_USER,
        Permission.UPDATE_ANY_USER,
        Permission.DELETE_USER,
        Permission.DELETE_ANY_USER,
        Permission.ASSIGN_ROLE,
        Permission.SUSPEND_USER,
        
        Permission.CREATE_COURSE,
        Permission.READ_COURSE,
        Permission.READ_ALL_COURSES,
        Permission.UPDATE_COURSE,
        Permission.UPDATE_ANY_COURSE,
        Permission.DELETE_COURSE,
        Permission.DELETE_ANY_COURSE,
        Permission.PUBLISH_COURSE,
        Permission.FEATURE_COURSE,
        Permission.HIDE_COURSE,
        
        Permission.UPLOAD_CONTENT,
        Permission.UPDATE_CONTENT,
        Permission.DELETE_CONTENT,
        Permission.MODERATE_CONTENT,
        
        Permission.ENROLL_COURSE,
        Permission.VIEW_ENROLLMENTS,
        Permission.VIEW_ALL_ENROLLMENTS,
        Permission.MANAGE_ENROLLMENTS,
        
        Permission.VIEW_PROGRESS,
        Permission.VIEW_ALL_PROGRESS,
        Permission.UPDATE_PROGRESS,
        Permission.TAKE_QUIZ,
        Permission.CREATE_NOTE,
        
        Permission.ASK_QUESTION,
        Permission.ANSWER_QUESTION,
        Permission.MODERATE_QA,
        
        Permission.VIEW_ANALYTICS,
        Permission.VIEW_OWN_ANALYTICS,
        Permission.VIEW_COURSE_ANALYTICS,
        Permission.VIEW_PLATFORM_ANALYTICS,
        Permission.EXPORT_ANALYTICS,
        
        Permission.PROCESS_PAYMENTS,
        Permission.VIEW_TRANSACTIONS,
        Permission.VIEW_ALL_TRANSACTIONS,
        Permission.MANAGE_PAYOUTS,
        Permission.VIEW_REVENUE,
        
        Permission.MANAGE_CATEGORIES,
        Permission.MANAGE_SETTINGS,
        Permission.MANAGE_EMAIL_TEMPLATES,
        Permission.VIEW_SYSTEM_HEALTH,
        Permission.MANAGE_LEGAL_PAGES,
        
        Permission.ISSUE_CERTIFICATE,
        Permission.VIEW_CERTIFICATES,
        Permission.VERIFY_CERTIFICATE,
    },
    
    UserRole.INSTRUCTOR: {
        # Instructor permissions
        Permission.READ_USER,  # Can read basic user info
        
        Permission.CREATE_COURSE,
        Permission.READ_COURSE,
        Permission.UPDATE_COURSE,  # Own courses only
        Permission.DELETE_COURSE,  # Own courses only
        Permission.PUBLISH_COURSE,  # Own courses only
        
        Permission.UPLOAD_CONTENT,
        Permission.UPDATE_CONTENT,
        Permission.DELETE_CONTENT,
        
        Permission.ENROLL_COURSE,  # Can enroll in other courses
        Permission.VIEW_ENROLLMENTS,  # Own course enrollments
        
        Permission.VIEW_PROGRESS,  # Student progress in own courses
        Permission.TAKE_QUIZ,
        Permission.CREATE_NOTE,
        
        Permission.ASK_QUESTION,
        Permission.ANSWER_QUESTION,  # In own courses
        
        Permission.VIEW_OWN_ANALYTICS,
        Permission.VIEW_COURSE_ANALYTICS,  # Own courses only
        
        Permission.VIEW_TRANSACTIONS,  # Own transactions
        Permission.VIEW_REVENUE,  # Own revenue
        
        Permission.ISSUE_CERTIFICATE,  # For own courses
        Permission.VIEW_CERTIFICATES,
        Permission.VERIFY_CERTIFICATE,
    },
    
    UserRole.LEARNER: {
        # Learner permissions
        Permission.READ_USER,  # Own profile only
        Permission.UPDATE_USER,  # Own profile only
        
        Permission.READ_COURSE,  # Public courses
        
        Permission.ENROLL_COURSE,
        Permission.VIEW_ENROLLMENTS,  # Own enrollments
        
        Permission.VIEW_PROGRESS,  # Own progress
        Permission.TAKE_QUIZ,
        Permission.CREATE_NOTE,
        
        Permission.ASK_QUESTION,
        
        Permission.VIEW_CERTIFICATES,  # Own certificates
        Permission.VERIFY_CERTIFICATE,
    }
}


class PermissionChecker:
    """
    Utility class for checking user permissions.
    """
    
    @staticmethod
    def has_permission(user_role: UserRole, permission: Permission) -> bool:
        """
        Check if a user role has a specific permission.
        
        Args:
            user_role: The user's role
            permission: The permission to check
            
        Returns:
            bool: True if the role has the permission, False otherwise
        """
        role_permissions = ROLE_PERMISSIONS.get(user_role, set())
        return permission in role_permissions
    
    @staticmethod
    def get_role_permissions(user_role: UserRole) -> Set[Permission]:
        """
        Get all permissions for a specific role.
        
        Args:
            user_role: The user's role
            
        Returns:
            Set[Permission]: Set of permissions for the role
        """
        return ROLE_PERMISSIONS.get(user_role, set())
    
    @staticmethod
    def can_access_resource(user_role: UserRole, required_permissions: List[Permission]) -> bool:
        """
        Check if a user role can access a resource that requires specific permissions.
        
        Args:
            user_role: The user's role
            required_permissions: List of permissions required for the resource
            
        Returns:
            bool: True if the role has at least one of the required permissions
        """
        role_permissions = ROLE_PERMISSIONS.get(user_role, set())
        return any(permission in role_permissions for permission in required_permissions)
    
    @staticmethod
    def requires_all_permissions(user_role: UserRole, required_permissions: List[Permission]) -> bool:
        """
        Check if a user role has all the required permissions.
        
        Args:
            user_role: The user's role
            required_permissions: List of permissions that are all required
            
        Returns:
            bool: True if the role has all required permissions
        """
        role_permissions = ROLE_PERMISSIONS.get(user_role, set())
        return all(permission in role_permissions for permission in required_permissions)

File: frontend/app/utils/permissions.ts
/**
 * Comprehensive role-based access control system
 * Defines all permissions and role hierarchies for the LMS
 */

export type UserRole = "super_admin" | "instructor" | "learner";

export type Permission =
    // Course Management
    | "create_course"
    | "update_course"
    | "delete_course"
    | "publish_course"
    | "view_all_courses"
    | "manage_course_content"

    // Content Management
    | "upload_content"
    | "update_content"
    | "delete_content"
    | "manage_media"

    // User Management
    | "create_user"
    | "update_user"
    | "delete_user"
    | "view_all_users"
    | "manage_user_roles"
    | "suspend_user"
    | "activate_user"

    // Enrollment Management
    | "enroll_user"
    | "unenroll_user"
    | "view_enrollments"
    | "manage_enrollments"

    // Analytics & Reporting
    | "view_system_analytics"
    | "view_course_analytics"
    | "view_user_analytics"
    | "view_financial_reports"
    | "export_reports"

    // Certificate Management
    | "issue_certificate"
    | "revoke_certificate"
    | "view_all_certificates"
    | "manage_certificate_templates"

    // Communication
    | "send_announcements"
    | "moderate_discussions"
    | "answer_questions"
    | "send_messages"

    // System Administration
    | "manage_system_settings"
    | "manage_themes"
    | "view_system_logs"
    | "manage_backups"
    | "manage_integrations"

    // Financial Management
    | "view_revenue"
    | "manage_pricing"
    | "process_refunds"
    | "view_transactions"

    // Instructor Management
    | "approve_instructors"
    | "review_instructor_applications"
    | "manage_instructor_payouts"

    // Learning Management
    | "take_quiz"
    | "view_progress"
    | "create_notes"
    | "ask_questions"
    | "rate_courses"
    | "view_certificates";

/**
 * Role-based permission mapping
 * Super Admin has ALL permissions by default
 */
export const ROLE_PERMISSIONS: Record<UserRole, Permission[]> = {
    super_admin: [], // Empty array means ALL permissions (handled in hasPermission function)

    instructor: [
        // Course Management
        "create_course",
        "update_course",
        "delete_course",
        "publish_course",
        "manage_course_content",

        // Content Management
        "upload_content",
        "update_content",
        "delete_content",
        "manage_media",

        // Analytics (own courses only)
        "view_course_analytics",

        // Enrollment (own courses only)
        "view_enrollments",

        // Certificate Management (own courses only)
        "issue_certificate",

        // Communication
        "answer_questions",
        "send_messages",
        "moderate_discussions",
    ],

    learner: [
        // Learning Activities
        "take_quiz",
        "view_progress",
        "create_notes",
        "ask_questions",
        "rate_courses",

        // Certificate Management (own certificates only)
        "view_certificates",
    ],
};

/**
 * Role hierarchy - higher roles inherit permissions from lower roles
 */
export const ROLE_HIERARCHY: Record<UserRole, UserRole[]> = {
    super_admin: ["instructor", "learner"], // Inherits all permissions
    instructor: ["learner"], // Inherits learner permissions
    learner: [], // Base role
};

/**
 * Special permissions that require additional context checks
 */
export const CONTEXTUAL_PERMISSIONS = {
    // These permissions require checking if user owns/manages the resource
    OWN_RESOURCE: [
        "update_course",
        "delete_course",
        "view_course_analytics",
        "issue_certificate",
    ],

    // These permissions are restricted to specific contexts
    ADMIN_ONLY: [
        "manage_system_settings",
        "view_system_logs",
        "manage_user_roles",
        "approve_instructors",
        "view_financial_reports",
    ],
};

/**
 * Check if a user has a specific permission
 */
export function hasPermission(
    userRole: UserRole,
    permission: Permission,
    context?: {
        resourceOwnerId?: number;
        userId?: number;
        isSystemAdmin?: boolean;
    }
): boolean {
    // Super admin has ALL permissions
    if (userRole === "super_admin") {
        return true;
    }

    // Check direct role permissions
    const rolePermissions = ROLE_PERMISSIONS[userRole];
    if (rolePermissions.includes(permission)) {
        // For contextual permissions, check additional conditions
        if (CONTEXTUAL_PERMISSIONS.OWN_RESOURCE.includes(permission)) {
            return context?.resourceOwnerId === context?.userId;
        }
        return true;
    }

    // Check inherited permissions from role hierarchy
    const inheritedRoles = ROLE_HIERARCHY[userRole];
    for (const inheritedRole of inheritedRoles) {
        if (ROLE_PERMISSIONS[inheritedRole].includes(permission)) {
            return true;
        }
    }

    return false;
}

/**
 * Check if a user has any of the specified permissions
 */
export function hasAnyPermission(
    userRole: UserRole,
    permissions: Permission[],
    context?: Parameters<typeof hasPermission>[2]
): boolean {
    return permissions.some(permission =>
        hasPermission(userRole, permission, context)
    );
}

/**
 * Check if a user has all of the specified permissions
 */
export function hasAllPermissions(
    userRole: UserRole,
    permissions: Permission[],
    context?: Parameters<typeof hasPermission>[2]
): boolean {
    return permissions.every(permission =>
        hasPermission(userRole, permission, context)
    );
}

/**
 * Get all permissions for a user role
 */
export function getAllPermissions(userRole: UserRole): Permission[] {
    // Super admin gets ALL permissions
    if (userRole === "super_admin") {
        return Object.values(ROLE_PERMISSIONS).flat().filter((permission, index, array) =>
            array.indexOf(permission) === index // Remove duplicates
        );
    }

    const permissions = new Set<Permission>();

    // Add direct permissions
    ROLE_PERMISSIONS[userRole].forEach(permission => permissions.add(permission));

    // Add inherited permissions
    const inheritedRoles = ROLE_HIERARCHY[userRole];
    inheritedRoles.forEach(role => {
        ROLE_PERMISSIONS[role].forEach(permission => permissions.add(permission));
    });

    return Array.from(permissions);
}

/**
 * Check if a role can access a specific route
 */
export function canAccessRoute(userRole: UserRole, route: string): boolean {
    // Super admin can access everything
    if (userRole === "super_admin") {
        return true;
    }

    // Route-based access control
    const routePermissions: Record<string, Permission[]> = {
        "/admin": ["manage_system_settings"],
        "/admin/users": ["view_all_users"],
        "/admin/courses": ["view_all_courses"],
        "/admin/analytics": ["view_system_analytics"],
        "/admin/financial": ["view_financial_reports"],
        "/admin/instructor-applications": ["approve_instructors"],
        "/admin/theme": ["manage_themes"],
        "/instructor": ["create_course"],
        "/instructor/courses": ["create_course"],
        "/instructor/analytics": ["view_course_analytics"],
        "/dashboard": [], // All authenticated users
        "/certificates": ["view_certificates"],
    };

    const requiredPermissions = routePermissions[route];
    if (!requiredPermissions) {
        return true; // No specific permissions required
    }

    if (requiredPermissions.length === 0) {
        return true; // Route accessible to all authenticated users
    }

    return hasAnyPermission(userRole, requiredPermissions);
}

/**
 * Get user-friendly role display name
 */
export function getRoleDisplayName(role: UserRole): string {
    const roleNames: Record<UserRole, string> = {
        super_admin: "Super Administrator",
        instructor: "Instructor",
        learner: "Learner",
    };

    return roleNames[role] || role;
}

/**
 * Get role badge color for UI display
 */
export function getRoleBadgeColor(role: UserRole): string {
    const colors: Record<UserRole, string> = {
        super_admin: "bg-red-100 text-red-800",
        instructor: "bg-blue-100 text-blue-800",
        learner: "bg-green-100 text-green-800",
    };

    return colors[role] || "bg-gray-100 text-gray-800";
}

/**
 * Navigation items with permission requirements
 */
export interface NavItemPermission {
    label: string;
    href: string;
    permissions?: Permission[];
    roles?: UserRole[];
    icon?: string;
    badge?: string;
    children?: NavItemPermission[];
}

/**
 * Get filtered navigation items based on user permissions
 */
export function getAuthorizedNavItems(
    userRole: UserRole,
    navItems: NavItemPermission[]
): NavItemPermission[] {
    return navItems.filter(item => {
        // Check role requirements
        if (item.roles && !item.roles.includes(userRole) && userRole !== "super_admin") {
            return false;
        }

        // Check permission requirements
        if (item.permissions && !hasAnyPermission(userRole, item.permissions)) {
            return false;
        }

        return true;
    }).map(item => ({
        ...item,
        children: item.children ? getAuthorizedNavItems(userRole, item.children) : undefined,
    }));
}